src/material/list/testing/list-harness.ts

```diff
@@ -0,0 +1,299 @@
+import {
+  MatActionListItemHarness,
+  MatListItemHarness,
+  MatListOptionHarness,
+  MatNavListItemHarness,
+  MatSubheaderHarness
+} from './list-item-harness';
+
+/** Represents a section of a list falling under a specific header. */
+export interface ListSection<I> {
+  /** The heading for this list section. `undefined` if there is no heading. */
+  heading?: string;
+
+  /** The items in this list section. */
+  items: I[];
+}
+
+/**
+ * Shared behavior among the harnesses for the various `MatList` flavors.
+ * @docs-private
+ */
+export class MatListHarnessBase
+    <
+      T extends (ComponentHarnessConstructor<C> & {
+        with: (options?: BaseHarnessFilters) => HarnessPredicate<C>
+      }),
+      C extends ComponentHarness,
+      F extends BaseListItemHarnessFilters
+    > extends ComponentHarness {
+  protected _itemHarness: T;
+
+  /**
+   * Gets a list of harnesses representing the items in this list.
+   * @param filters Optional filters used to narrow which harnesses are included
+   * @return The list of items matching the given filters.
+   */
+  async getItems(filters?: F): Promise<C[]> {
+    return this.locatorForAll(this._itemHarness.with(filters))();
+  }
+
+  /**
+   * Gets a list of `ListSection` representing the list items grouped by subheaders. If the list has
+   * no subheaders it is represented as a single `ListSection` with an undefined `heading` property.
+   * @param filters Optional filters used to narrow which list item harnesses are included
+   * @return The list of items matching the given filters, grouped into sections by subheader.
+   */
+  async getItemsBySubheader(filters?: F): Promise<ListSection<C>[]> {
+    const listSections = [];
+    let currentSection: ListSection<C> = {items: []};
+    const itemsAndSubheaders =
+        await this.getItemsSubheadersAndDividers({item: filters, divider: false});
+    for (const itemOrSubheader of itemsAndSubheaders) {
+      if (itemOrSubheader instanceof MatSubheaderHarness) {
+        if (currentSection.heading !== undefined || currentSection.items.length) {
+          listSections.push(currentSection);
+        }
+        currentSection = {heading: await itemOrSubheader.getText(), items: []};
+      } else {
+        currentSection.items.push(itemOrSubheader);
+      }
+    }
+    if (currentSection.heading !== undefined || currentSection.items.length ||
+        !listSections.length) {
+      listSections.push(currentSection);
+    }
+    return listSections;
+  }
+
+  /**
+   * Gets a list of sub-lists representing the list items grouped by dividers. If the list has no
+   * dividers it is represented as a list with a single sub-list.
+   * @param filters Optional filters used to narrow which list item harnesses are included
+   * @return The list of items matching the given filters, grouped into sub-lists by divider.
+   */
+  async getItemsDivided(filters?: F): Promise<C[][]> {
+    const listSections = [];
+    let currentSection = [];
```

Comment:
You could avoid having two arrays by pushing to `listSections[listSections.length - 1]`.